File: src/SourceGenerator/Trulla.SourceGenerator.Core/CodeGen/Text.fs
module Trulla.SourceGenerator.Text

open System.Text

type StringBuffer = StringBuilder -> unit

// Base from:
// http://fssnip.net/7WR
// https://github.com/fsharp/fslang-suggestions/issues/775
type TextBuilder () =
    member inline _.Yield (txt: string) = fun (b: StringBuilder) -> 
        Printf.bprintf b "%s" txt
    member inline _.Yield (c: char) = fun (b: StringBuilder) -> 
        Printf.bprintf b "%c" c
    member inline _.Yield (strings: #seq<string>) = fun (b: StringBuilder) ->
        for s in strings do Printf.bprintf b "%s\n" s
    member inline _.YieldFrom (f: StringBuffer) =
        f
    member inline _.Combine (f,g) = fun (b: StringBuilder) ->
        f b; g b
    member inline _.Delay (f) = fun (b: StringBuilder) ->
        (f()) b
    member inline _.Zero () =
        ignore
    member inline _.For (xs: 'a seq, f: 'a -> StringBuffer) = fun (b: StringBuilder) ->
        use e = xs.GetEnumerator ()
        while e.MoveNext() do
            (f e.Current) b
    member inline _.While (p: unit -> bool, f: StringBuffer) = fun (b: StringBuilder) ->
        while p () do f b
    member inline _.Run (f: StringBuffer) =
        let b = StringBuilder()
        do f b
        b.ToString()

let text = TextBuilder()

let quot = "\""
let plus = " + "
let pareno = "("
let parenc = ")"
let br = "\n"

let getIndentation indent = String.replicate indent "    "
let ind indent (txt: string) = text { getIndentation indent; txt }
let lni indent (txt: string) = ind indent (text { txt; br })
let ln0 = lni 0
let stri indent (txt: string) = ind indent (text { quot; txt; quot })
let str (txt: string) = stri 0 txt
let strlni indent (txt: string) = ln0 (stri indent txt)
let strln (txt: string) = ln0 (str txt)

File: src/NodeCli/lib/Common/CoreFiles/Text.js
import { class_type } from "../../fable_modules/fable-library.4.3.0/Reflection.js";
import { printf, replicate } from "../../fable_modules/fable-library.4.3.0/String.js";
import { StringBuilder_$ctor } from "../../fable_modules/fable-library.4.3.0/System.Text.js";
import { PrintfModule_PrintFormatToStringBuilder } from "../../fable_modules/fable-library.4.3.0/FSharp.Core.js";
import { toString } from "../../fable_modules/fable-library.4.3.0/Types.js";

export class TextBuilder {
    constructor() {
    }
}

export function TextBuilder_$reflection() {
    return class_type("Trulla.Core.Text.TextBuilder", void 0, TextBuilder);
}

export function TextBuilder_$ctor() {
    return new TextBuilder();
}

export const text = TextBuilder_$ctor();

export const quot = "\"";

export const plus = " + ";

export const pareno = "(";

export const parenc = ")";

export const br = "\n";

export function getIndentation(indent) {
    return replicate(indent, "    ");
}

export function ind(indent, txt) {
    let txt_1;
    const b_5 = StringBuilder_$ctor();
    const b_3 = b_5;
    ((txt_1 = getIndentation(indent), (b) => {
        PrintfModule_PrintFormatToStringBuilder(b, printf("%s"))(txt_1);
    }))(b_3);
    PrintfModule_PrintFormatToStringBuilder(b_3, printf("%s"))(txt);
    return toString(b_5);
}

export function lni(indent, txt) {
    let b_5, b_3;
    return ind(indent, (b_5 = StringBuilder_$ctor(), ((b_3 = b_5, (PrintfModule_PrintFormatToStringBuilder(b_3, printf("%s"))(txt), PrintfModule_PrintFormatToStringBuilder(b_3, printf("%s"))(br))), toString(b_5))));
}

export const ln0 = (txt) => lni(0, txt);

export function stri(indent, txt) {
    let b_8, b_6, b_4;
    return ind(indent, (b_8 = StringBuilder_$ctor(), ((b_6 = b_8, (PrintfModule_PrintFormatToStringBuilder(b_6, printf("%s"))(quot), (b_4 = b_6, (PrintfModule_PrintFormatToStringBuilder(b_4, printf("%s"))(txt), PrintfModule_PrintFormatToStringBuilder(b_4, printf("%s"))(quot))))), toString(b_8))));
}

export function str(txt) {
    return stri(0, txt);
}

export function strlni(indent, txt) {
    return ln0(stri(indent, txt));
}

export function strln(txt) {
    return ln0(str(txt));
}

